environment/src/main/java/com/sequenceiq/environment/environment/service/DatahubService.java

```
@@ -0,0 +1,98 @@
+
+    private final WebApplicationExceptionMessageExtractor webApplicationExceptionMessageExtractor;
+
+    public DatahubService(DistroXV1Endpoint distroXV1Endpoint, DatahubPollerProvider datahubPollerProvider,
+            WebApplicationExceptionMessageExtractor webApplicationExceptionMessageExtractor) {
+        this.distroXV1Endpoint = distroXV1Endpoint;
+        this.datahubPollerProvider = datahubPollerProvider;
+        this.webApplicationExceptionMessageExtractor = webApplicationExceptionMessageExtractor;
+    }
+
+    public void startAttachedDatahubClusters(Long envId, String envCrn) {
+        Collection<StackViewV4Response> attachedDistroXClusters = getAttachedDatahubClusters(envCrn);
+        List<String> datahubCrns = mapStacksToCrns(attachedDistroXClusters);
+        executeDatahubOperationAndStartPolling(datahubCrns, distroXV1Endpoint::putStartByCrns,
+                datahubPollerProvider.startDatahubClustersPoller(datahubCrns, envId));
+
+    }
+
+    public void stopAttachedDatahubClusters(Long envId, String envCrn) {
+        Collection<StackViewV4Response> attachedDistroXClusters = getAttachedDatahubClusters(envCrn);
+        List<String> datahubCrns = mapStacksToCrns(attachedDistroXClusters);
+        executeDatahubOperationAndStartPolling(datahubCrns, distroXV1Endpoint::putStopByCrns,
+                datahubPollerProvider.stopDatahubClustersPoller(datahubCrns, envId));
+    }
+
+    private void executeDatahubOperationAndStartPolling(List<String> datahubCrns, Consumer<List<String>> distroxOperation, AttemptMaker<Void> attemptMaker) {
+        try {
+            distroxOperation.accept(datahubCrns);
+            Polling.stopAfterAttempt(attemptCount)
+                    .stopIfException(true)
+                    .waitPeriodly(sleepTime, TimeUnit.SECONDS)
+                    .run(attemptMaker);
+        } catch (PollerException e) {
+            if (e.getCause() != null && e.getCause() instanceof WebApplicationException) {
```

otherwise you swallow exception. rethrow please! if no rethrow is needed please leave a comment in the code why is it normal.